### ISSUE TITLE
Handling cancellation within a BlockProcedure block

### ISSUE THREAD

**Issue by swiftlyfalling** [ProcedureKit/ProcedureKit#654]
Suppose someone wants to create a (synchronous) BlockProcedure, and also make it cancellable (i.e. check `!isCancelled` periodically inside the block).

Currently, they can create a BlockProcedure as follows:

```swift
let procedure = BlockProcedure {
    // do something
}
```

But they can't do something like this, because the variable `procedure` will be used within its own initial value:

```swift
let procedure = BlockProcedure {
    while !procedure.isCancelled { // 🚫 "Variable used within its own initial value"
        // do something
    }
}
```

Should we make this cleaner by passing-in the procedure to the block?

i.e. Modify BlockProcedure to take a block with type `(BlockProcedure) throws -> Output`?
Or provide a separate init that support this, like:

```swift
public init(cancellableBlock block: @escaping (ResultProcedure) throws -> Output) {
    self.block = block
    super.init()
}
```

That way, the solution becomes:

```swift
let procedure = BlockProcedure(cancellableBlock: { procedure in
    while !procedure.isCancelled { // works
        // do something
    }
})
```

**Comment by swiftlyfalling** [comment]
Alternatively, the cancellableBlock could provide a closure that returns whether the block should cancel.

i.e.

```swift
public init(cancellableBlock block: @escaping (() -> Bool) throws -> Output) {
    self.block = block
    super.init()
}
```

Use:

```swift
let procedure = BlockProcedure(cancellableBlock: { isCancelled in
    while !isCancelled() {
        // do something
    }
})
```

**Comment by swiftlyfalling** [comment]
I kinda like the second method. Thoughts?

```swift
open class CancellableResultProcedure<Output>: Procedure, OutputProcedure {

    public typealias ThrowingCancellableOutputBlock = (() -> Bool) throws -> Output

    public var output: Pending<ProcedureResult<Output>> = .pending

    private let block: ThrowingCancellableOutputBlock

    public init(cancellableBlock: @escaping ThrowingCancellableOutputBlock) {
        self.block = cancellableBlock
        super.init()
    }

    open override func execute() {
        defer { finish(withError: output.error) }
        do { output = .ready(.success(try block({ return isCancelled }))) }
        catch { output = .ready(.failure(error)) }
    }
}

open class CancellableBlockProcedure: CancellableResultProcedure<Void> { }
```

Use like this:

```swift
let procedure = CancellableBlockProcedure(cancellableBlock: { isCancelled in
    while !isCancelled() {
        // do something
    }
})
```

**Comment by swiftlyfalling** [comment]
This was implemented as part of #668, which has been merged into `development`.
